google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/ConnectionOptionsTest.java

```diff
@@ -727,116 +728,208 @@ public void testInvalidCredentials() {
+                  FILE_TEST_PATH, encoded);
 
-    IllegalArgumentException e =
-        assertThrows(
-            IllegalArgumentException.class,
-            () -> ConnectionOptions.newBuilder().setUri(uri).build());
-    assertTrue(
-        e.getMessage(),
-        e.getMessage()
-            .contains(
-                "Specify only one of credentialsUrl, encodedCredentials, credentialsProvider and OAuth token"));
+          IllegalArgumentException e =
+              assertThrows(
+                  IllegalArgumentException.class,
+                  () -> ConnectionOptions.newBuilder().setUri(uri).build());
+          assertTrue(
+              e.getMessage(),
+              e.getMessage()
+                  .contains(
+                      "Specify only one of credentialsUrl, encodedCredentials, credentialsProvider and OAuth token"));
+        });
   }
 
   public static class TestCredentialsProvider implements CredentialsProvider {
     @Override
-    public Credentials getCredentials() throws IOException {
+    public Credentials getCredentials() {
       return NoCredentials.getInstance();
     }
   }
 
+  static void runWithSystemPropertyEnabled(String systemPropertyName, ThrowingRunnable runnable)
+      throws Throwable {
+    String originalValue = System.getProperty(systemPropertyName);
+    System.setProperty(systemPropertyName, "true");
+    try {
+      runnable.run();
+    } finally {
+      if (originalValue == null) {
+        System.clearProperty(systemPropertyName);
+      } else {
+        System.setProperty(systemPropertyName, originalValue);
+      }
+    }
+  }
+
+  @Test
+  public void testValidCredentialsProvider() throws Throwable {
+    runWithSystemPropertyEnabled(
+        ConnectionOptions.ENABLE_CREDENTIALS_PROVIDER_SYSTEM_PROPERTY,
+        () -> {
+          String uri =
+              String.format(
+                  "cloudspanner:/projects/test-project/instances/test-instance/databases/test-database?credentialsProvider=%s",
+                  TestCredentialsProvider.class.getName());
+
+          ConnectionOptions options = ConnectionOptions.newBuilder().setUri(uri).build();
+          assertEquals(NoCredentials.getInstance(), options.getCredentials());
+        });
+  }
+
   @Test
-  public void testValidCredentialsProvider() {
+  public void testValidCredentialsProvider_WithoutEnablingSystemProperty() {
     String uri =
         String.format(
             "cloudspanner:/projects/test-project/instances/test-instance/databases/test-database?credentialsProvider=%s",
             TestCredentialsProvider.class.getName());
+    SpannerException exception =
+        assertThrows(
+            SpannerException.class, () -> ConnectionOptions.newBuilder().setUri(uri).build());
+    assertEquals(ErrorCode.FAILED_PRECONDITION, exception.getErrorCode());
+    assertEquals(
+        "FAILED_PRECONDITION: credentialsProvider can only be used if the system property ENABLE_CREDENTIALS_PROVIDER has been set to true. "
```

Comment:
The error that you get when setting a CredentialsProvider without enabling the property is verified here.